_source/_posts/2022-07-06-spa-web-security.md

```diff
@@ -0,0 +1,137 @@
+Check out a step-by-step guide on dependency management for JavaScript developers by the OWASP team - [NPM Security best practices](https://cheatsheetseries.owasp.org/cheatsheets/NPM_Security_Cheat_Sheet.html).
+
+## Load resources securely
+A particular concern for SPAs are cross-origin resources, that is, resources your web app uses that are not hosted on the same origin as your web application. When all the resources the front-end needs are available through the same origin, we have confidence that the resources our secure. Hopefully, we're not pwning ourselves! At least not intentionally, right?!
+
+A typical pattern for setting up URLs for SPA communicating to back-end APIs might look like the following, where your front-end is
+
+```
+https://myfavekdramas.com
+```
+
+and the back-end API URL is
+
+```
+https://apis.myfavekdramas.com
+```
+
+But the API URL is a subdomain, so the request is cross-origin. Most browsers (once again BUH-BYE Internet Explorer!) protect us by allowing only specific cross-origin requests. 
+
+Between your production environment setup looking like the above URLs and local development having different ports for your front-end and back-end, you might be tempted to enable CORS to allow all sorts of requests. Unlike other security measures covered in this post, your browser already protects you in the strictest way possible. Enabling CORS _loosens_ that restriction.
+
+So the advice is to enable CORS appropriately with considered use of allowlists and to use proxies as appropriate (such as for local development). The concept of least privilege applies here too, and keeping a tight lid on things helps mitigate security vulnerabilities.
+
+### Dive into cross-origin resource sharing
+Want to understand better when CORS rules apply or how to set up the allowlist rules? Read these resources for a deep dive.
+* [MDN CORS Reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS)
+* [Auth0's CORS Tutorial](https://auth0.com/blog/cors-tutorial-a-guide-to-cross-origin-resource-sharing/)
+
+## Protect your cookies from theft
+Really, this advice is applicable outside of web development too. But since we're talking about web development in this post, we'll skip mitigation strategies about protecting your physical cookies from marauders and focus again on security headers.
+
+Cookies can contain all sorts of goodies, such as chocolate chips and the tidbits of data that agitators love to have! Current authentication best practices favor using in-browser memory or web workers for authentication tokens over cookies, and cookies should never be used for secrets, but there might still be sensitive information stored in cookies. To help protect cookies, we can use attributes.
+
+Cookies should be temporary. Particularly cookies with sensitive information should be as short-lived as possible. When you or your back-end bakes up a cookie, adding the attribute `Max-Age` defines the longevity of the cookie. 
```

Comment:
```suggestion
Cookies should be temporary. In particular, cookies with sensitive information should be as short-lived as possible. When you or your back-end bakes up a cookie, adding the attribute `Max-Age` defines the longevity of the cookie. 
```
use the transition phrase instead of an adverb